consensus_encoding/src/encode/encoders.rs

```diff
@@ -76,6 +76,88 @@ impl<const N: usize> Encoder<'_> for ArrayEncoder<N> {
     }
 }
 
+/// An encoder for a list of encodable types.
+pub struct SliceEncoder<'e, T: Encodable> {
+    /// The list of references to the objects we are encoding.
+    ///
+    /// This is **never** mutated. All accesses are done by array accesses because
+    /// of lifetimes and the borrow checker.
+    sl: &'e [T],
+    /// Length of `sl`.
+    len: usize,
+    /// `len` encoded as a compact size.
+    compact_size: Option<ArrayVec<u8, SIZE>>,
+    /// Index into `sl` of the element we are currently encoding.
+    cur_idx: usize,
+    /// Current encoder (for `sl[self.cur_idx]`).
+    cur_enc: Option<T::Encoder<'e>>,
+}
+
+impl<'e, T: Encodable> SliceEncoder<'e, T> {
+    /// Constructs an encoder which encodes the slice with a length prefix.
+    pub fn with_length_prefix(sl: &'e [T]) -> Self {
+        let len = sl.len();
+        let compact_size = Some(compact_size::encode(len));
+
+        if sl.is_empty() {
+            Self { sl, len: 0, compact_size, cur_idx: 0, cur_enc: None }
+        } else {
+            let encoder = sl[0].encoder();
+            Self { sl, len, compact_size, cur_idx: 0, cur_enc: Some(encoder) }
```

Comment:
In cd866ab3994bed858d11bc3ba33e0aada50dd3fb:

nit: I would've just done `cur_enc: sl.get(0).map(T::encoder)` which would avoid the if statement and be clearer.

Reply:
This is messing with my mind, but here is a theory. The `T::encoder` method creates an encoder which requires a reference to the type `T` which it is encoding. In this `map`, that references is being derived from the short lived reference passed as an arg to the `map` function. I want to learn the technical name for it, but I think a reference derived from another reference is bound by to the lifetime of the reference, not the underlying data it points to. Maybe it is as simple as a reference to a reference?

Reply:
This is extremely weird. `sl.get(0).map(T::encoder) ` does not work but `sl.get(0).map(|x| x.encoder())` does, even though the types match up.

Reply:
Self-contained reproduction (will file an issue if I don't forget)

```rust
trait Encodable {
    type Encoder<'e> where Self: 'e;
    fn encoder(&self) -> Self::Encoder<'_>;
}

impl Encodable for u32 {
    type Encoder<'e> = &'e u32;
    
    fn encoder(&self) -> Self::Encoder<'_> { self }
}


struct RefHolder<'e, T: Encodable + 'e> {
    rf: Option<T::Encoder<'e>>,
}

fn ref_holder<'e, T: Encodable + 'e>(sl: &'e [T]) -> RefHolder<'e, T> {
    RefHolder {
        rf: sl.get(0).map(T::encoder),
    }
}

fn main() {}
```

Reply:
Quite the rabbit hole. It seems like it has something to do with the compiler struggling to connect the dots with the function item version (`map(T::encoder)`) due to that being backed by a Higher Ranked Trait Bound. I guess the closure version is implemented without them and the compiler can trace the lifetime. But I haven't found any satisfying docs or specific issues on it yet...